compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3840,9 +3840,15 @@ class Typer extends Namer
 
         // approximate type params with bounds
         def approx = new ApproximatingTypeMap {
+          var alreadyExpanding: List[TypeRef] = Nil
           def apply(tp: Type) = tp.dealias match
             case tp: TypeRef if !tp.symbol.isClass =>
-              expandBounds(tp.info.bounds)
+              if alreadyExpanding contains tp then tp else
```

Comment:
How is this connected to the rest of the changes? Did the rest trigger a SO here?

Reply:
Yes, this is the source of the SO that happened after restoring GADT constraints on failed subtype checks.